nemo/collections/asr/parts/submodules/spectr_augment.py

```diff
@@ -142,40 +142,43 @@ def _forward_fast(self, input_spec: torch.Tensor, length: torch.Tensor) -> torch
     def _apply_masks(
             self, input_spec: torch.Tensor, num_masks: int, length: torch.Tensor, width: int | float, 
             mask_value: float, axis: int) -> torch.Tensor:
+
         batch_size = input_spec.shape[0]
         axis_length = input_spec.shape[axis]
 
-        # We use float32 dtype for begin/end mask markers before they are quantized to long.
-        # Using x.dtype might cause us to encounter dtypes such as bf16 or smaller which
-        # wouldn't be able to represent every frame index leading to subtle bugs.
+        #If width is float then it is transformed into a tensor
         if isinstance(width, float):
-            scaled_width = torch.clamp(width * length, max = axis_length).unsqueeze(1)
-            mask_width =  torch.rand((batch_size, num_masks), device=input_spec.device, dtype=torch.float32) * scaled_width
-            mask_width = mask_width.long()
-            mask_start = torch.rand((batch_size, num_masks), device=input_spec.device, dtype=torch.float32) * (axis_length - mask_width)
-            mask_start.long()
-        else:
-            #Since we don't need to compute scaled width, we can call randint mask_width and mask_start
-            width = min(width, axis_length)
-            mask_start = torch.randint(low=0, high = max(1, axis_length - width), size=(batch_size, num_masks), 
-                device=input_spec.device, dtype=torch.long)
-            mask_width = torch.randint(low=0, high = max(1, width), size=(batch_size, num_masks), 
-                device=input_spec.device, dtype=torch.long)
-            mask_end = mask_start + mask_width
+                width = torch.clamp(width * length, max = axis_length).unsqueeze(1)
 
+        # Generate [0-1) random numbers and then scale the tensors.
+        # Use float32 dtype for begin/end mask markers before they are quantized to long.
+        # Using x.dtype might cause us to encounter dtypes such as bf16 or smaller which
+        # wouldn't be able to represent every frame index leading to subtle bugs.
+        mask_width = torch.rand((batch_size, num_masks), device=input_spec.device, dtype=torch.float32) * width
+        mask_width = mask_width.long()
+        mask_start = torch.rand((batch_size, num_masks), device=input_spec.device, dtype=torch.float32) * (axis_length - mask_width)
+        mask_start.long()
```

Comment:
should we assign the result? `mask_start = mask_start.long()`

Reply:
for reference the version that I tested and confirmed that it masks properly (ie skipping the padding regions) had the following code in this bit:

```python
if isinstance(width, float):
            width = torch.min(width * length, torch.tensor(axis_length, dtype=torch.float32)).unsqueeze(1)
        mask_width = torch.rand((batch_size, num_masks), device=input_spec.device, dtype=torch.float32) * width
        mask_width = mask_width.long()
        mask_start = torch.rand((batch_size, num_masks), device=input_spec.device, dtype=torch.float32) * (length.unsqueeze(1) - mask_width)
        mask_start = mask_start.long()
```

I see you are using axis_length rather than length for mask_start so I'm suspecting this version has the padding mask issue.

Reply:
@pzelasko you are right let me change it to length

Reply:
I fixed both